File: bubble-sort/bubble_sort.py
#Implementation of bubble sort in python.

def bubble_sorted(items):
    """creates a new list, sorts it using bubble_sort,
    then returns the new sorted list. The original list
    is left untouched."""
    from copy import deepcopy

    new = deepcopy(items)
    bubble_sort(new) #new is modified
    return new
    
def bubble_sort(items):
    """sorts the given list in place"""
    is_sorted = False
    while not is_sorted:
        is_sorted = True
        for i in range(len(items) - 1):
            if items[i] > items[i + 1]:
                items[i], items[i + 1] = items[i + 1], items[i]
                is_sorted = False
    return items #items is modified

if __name__ == "__main__":
    #executes this piece of code if the user runs this script directly,
    #i.e. not by importing.

    print("enter a space seperated list of numbers like 5 4 73 1.")
    item_list = bubble_sort(list(map(int, input("integers to be sorted: ").split())))

    #print(*[1,2,3]) is equivalent to print(1, 2, 3)
    print(*item_list)

File: bubble-sort/bubble_sort.c
#include <stdio.h>

/* bubble sort written in C */

void bubble_sort(float **arr, int size);

int main(void)
{
	/* driver code here */
	float *fp, arr[]={1,5,4,3,6,5,4,7,3,25,4,37,546,0,-15,0.25};
	int size = 16, j;
	fp = arr;

	printf("{");
	for (j=0; j<size-1; j++)
	{
		printf("%f, ", fp[j]);
	}
	printf("%f}\n", fp[size-1]);

	bubble_sort(&fp, size);

	printf("{");
	for (j=0; j<size-1; j++)
	{
		printf("%f, ", fp[j]);
	}
	printf("%f}\n", fp[size-1]);


	return 0;
}

void bubble_sort(float **arr, int size)
{
	/* 
	 * sorts in place.
	 * pass a pointer to an array of integers
	 * or basically a pointer to a float
	 * pointer to the function.
	 * */

	int is_sorted = 0, i, tmp;
	while (!is_sorted)
	{
		is_sorted = 1;
		for (i=0; i<size-1; i++)
		{
			if ((*arr)[i] > (*arr)[i+1])
			{
				tmp = (*arr)[i];
				(*arr)[i] = (*arr)[i+1];
				(*arr)[i+1] = tmp;
				is_sorted = 0;
			}
		}
	}
}
